src/ddmd/dmodule.d

```diff
@@ -922,44 +891,60 @@ extern (C++) final class Module : Package
                 members.append(p.parseDeclDefs(0));
             }
         }
-        // Insert module into the symbol table
-        Dsymbol s = this;
-        if (isPackageFile)
+    }
+
+    /**
+       Insert module name into package tree.
+       Package modules are wrapped inside a Package before insertion.
+       In case of conflicting modules the existing entry is returned.
+
+       Returns:
+         This module, the containing package module, or a previously inserted
+         conflicting module.
+     */
+    final Package insertIntoPackageTree()
+    {
+        DsymbolTable dst;
+        // Insert module name into symbol table
+        if (md)
         {
-            /* If the source tree is as follows:
-             *     pkg/
-             *     +- package.d
-             *     +- common.d
-             * the 'pkg' will be incorporated to the internal package tree in two ways:
-             *     import pkg;
-             * and:
-             *     import pkg.common;
-             *
-             * If both are used in one compilation, 'pkg' as a module (== pkg/package.d)
-             * and a package name 'pkg' will conflict each other.
-             *
-             * To avoid the conflict:
-             * 1. If preceding package name insertion had occurred by Package::resolve,
-             *    later package.d loading will change Package::isPkgMod to PKGmodule and set Package::mod.
-             * 2. Otherwise, 'package.d' wrapped by 'Package' is inserted to the internal tree in here.
+            /* A ModuleDeclaration, md, was provided.
+             * The ModuleDeclaration sets the packages this module appears in, and
+             * the name of this module.
              */
-            auto p = new Package(ident);
-            p.parent = this.parent;
-            p.isPkgMod = PKGmodule;
-            p.mod = this;
-            p.tag = this.tag; // reuse the same package tag
-            p.symtab = new DsymbolTable();
-            s = p;
+            Package ppack = null;
+            dst = Package.resolve(md.packages, &this.parent, &ppack);
+            assert(dst);
+            Module m = ppack ? ppack.isModule() : null;
+            if (m && (strcmp(m.srcfile.name.name(), "package.d") != 0 &&
+                      strcmp(m.srcfile.name.name(), "package.di") != 0))
+            {
+                .error(md.loc, "package name '%s' conflicts with usage as a module name in file %s", ppack.toPrettyChars(), m.srcfile.toChars());
+            }
         }
-        if (!dst.insert(s))
+        else
+        {
+            /* The name of the module is set to the source file name.
+             * There are no packages.
+             */
+            dst = Module.modules; // and so this module goes into global module symbol table
+            /* Check to see if module name is a valid identifier
+             */
+            if (!Identifier.isValidIdentifier(ident.toChars()))
```

Comment:
Mmh, looks like this now triggers for `spec/hash-map.dd`, will need to debug this.

Reply:
Forgot to check `isDocFile` in insertIntoPackageTree, should work now.